backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -891,63 +933,59 @@ public FutureExperimentRunDAO(
-                        .addActions(
-                            Action.newBuilder()
-                                .setModeldbServiceAction(action)
-                                .setService(ServiceEnum.Service.MODELDB_SERVICE))
-                        .addResources(
-                            Resources.newBuilder()
-                                .setService(ServiceEnum.Service.MODELDB_SERVICE)
-                                .setResourceType(
-                                    ResourceType.newBuilder()
-                                        .setModeldbServiceResourceType(
-                                            ModelDBServiceResourceTypes.REPOSITORY))
-                                .addAllResourceIds(repoId))
-                        .build()),
-            executor)
-        .thenCompose(
-            response -> InternalFuture.completedInternalFuture(response.getAllowed()), executor);
+  private InternalFuture<Map<String, VersioningEntry>> filterVersionedInputsBasedOnPrivileges(
+      Set<String> runIds, InternalFuture<Map<String, VersioningEntry>> futureVersionedInputs) {
+    return futureVersionedInputs.thenCompose(
+        versionedInputsMap -> {
+          List<InternalFuture<Map<String, VersioningEntry>>> internalFutureList = new ArrayList<>();
+          for (String runId : runIds) {
+            // Get VersionEntry from fetched map
+            VersioningEntry versioningEntry = versionedInputsMap.get(runId);
+            if (versioningEntry == null) {
+              continue;
+            }
+            String repoId = String.valueOf(versioningEntry.getRepositoryId());
+            // Check repository is accessible or not from versionedInputs If not then we will remove
+            // it from fetched map
+            internalFutureList.add(
+                InternalFuture.completedInternalFuture(versioningEntry)
+                    .thenCompose(
+                        versioningEntry1 ->
+                            checkEntityPermissionBasedOnResourceTypes(
+                                    Collections.singletonList(repoId),
+                                    ModelDBActionEnum.ModelDBServiceActions.READ,
+                                    ModelDBServiceResourceTypes.REPOSITORY)
+                                .thenCompose(
+                                    isSelfAllowed -> {
+                                      // Set null into map if repository is not accessible to the
+                                      // current user
+                                      VersioningEntry allowedVersioningEntry = null;
+                                      if (isSelfAllowed) {
+                                        allowedVersioningEntry = versioningEntry1;
+                                      }
+                                      return InternalFuture.completedInternalFuture(
+                                          Collections.singletonMap(runId, allowedVersioningEntry));
```

Comment:
Is it safe for us to create a map with `null` here? I'd probably expect this to be skipped vs saving a `null` entry as it will probably cause a NPE down the line.

Reply:
We can not just skip the entry because we have to return something so now i have returned empty entry instead of add the null in map.